src/com/luissolano/Main.java

```diff
@@ -203,6 +216,16 @@ void drain() {
                     }
                 }
             }
+
+            @Override
+            public void request(long l) {
+                s.request(l);
+            }
+
+            @Override
+            public void cancel() {
+                s.cancel();
+            }
```

Comment:
you have to dispose the worker in cancel and after any terminal event inside drain(). The requesting can stay as is here. Replenishment call(s) are needed from drain() though.